### PR TITLE
Mocha add promise

### PR DESCRIPTION
**Pull request by frogcjn**
mocha originally supports promise with action function.
https://mochajs.org/#asynchronous-code

So this code

``` ts
it("test hello", async function () {
            await request(app).post("/hello")
                .expect("Content-Type", /json/)
                .expect(success)
        })
```

is possible in TypeScript.
(This code is already possible, because mocha supports promise.

### CODE REVIEW

File: mocha/mocha.d.ts [hunk 1]
```diff
@@ -46,55 +42,29 @@ declare var xit: Mocha.ITestDefinition;
...
-declare function afterEach(description: string, action: () => void): void;
-
-declare function afterEach(description: string, action: (done: MochaDone) => void): void;
+interface MochaDone {
+    (error?: any): any;
+}
 
-declare function suiteTeardown(action: () => void): void;
+interface ActionFunction {
+    (done?: MochaDone): any | PromiseLike<any>
```
Comment (mhegazy) [inline mocha/mocha.d.ts]: do not use optional arguments in call back. the caller will call with a value all the time. the callee can decide to ignore it or not. but making it optional does not really mean that.

File: mocha/mocha.d.ts [hunk 2]
```diff
@@ -46,55 +42,29 @@ declare var xit: Mocha.ITestDefinition;
 declare var test: Mocha.ITestDefinition;
 declare var specify: Mocha.ITestDefinition;
 
-declare function before(action: () => void): void;
```
Comment (mhegazy) [inline mocha/mocha.d.ts]: why were these removed?

File: mocha/mocha.d.ts [hunk 3]
```diff
@@ -46,55 +42,29 @@ declare var xit: Mocha.ITestDefinition;
...
+declare function suiteSetup(action: ActionFunction): void;
+declare function suiteTeardown(action: ActionFunction): void;
+declare function before(action: ActionFunction): void;
+declare function before(description: string, action: ActionFunction): void;
+declare function after(action: ActionFunction): void;
+declare function after(description: string, action: ActionFunction): void;
+declare function beforeEach(action: ActionFunction): void;
+declare function beforeEach(description: string, action: ActionFunction): void;
+declare function afterEach(action: ActionFunction): void;
+declare function afterEach(description: string, action: ActionFunction): void;
```
Comment (frogcjn) [inline mocha/mocha.d.ts]: @mhegazy  it is not be removed. it is modified here.

**Comment by frogcjn** [comment]
@mhegazy , here is why there should be a optional.

File: mocha/mocha.d.ts (hunk 1 above)
Comment (frogcjn) [inline mocha/mocha.d.ts]: @mhegazy no optional function type will not match signature. I have tried, and you can see the output of CI test.

**Comment by frogcjn** [comment]
@mhegazy Thanks!
